src/beginner/ros/wheel-encoder-reader.md

```diff
@@ -0,0 +1,156 @@
+        # static parameters
+        self._vehicle_name = os.environ['VEHICLE_NAME']
+        self._left_encoder_topic = f"/{self._vehicle_name}/left_wheel_encoder_node/tick"
+        self._right_encoder_topic = f"/{self._vehicle_name}/right_wheel_encoder_node/tick"
+        # temporary data storage
+        self._ticks_left = None
+        self._ticks_right = None
+        # construct subscriber
+        self.sub_left = rospy.Subscriber(self._left_encoder_topic, WheelEncoderStamped, self.callback_left)
+        self.sub_right = rospy.Subscriber(self._right_encoder_topic, WheelEncoderStamped, self.callback_right)
+
+    def callback_left(self, data):
+        # log general information once at the beginning
+        rospy.loginfo_once(f"Left encoder resolution: {data.resolution}")
+        rospy.loginfo_once(f"Left encoder type: {data.type}")
+        # store data value
+        self._ticks_left = data.data
+
+    def callback_right(self, data):
+        # log general information once at the beginning
+        rospy.loginfo_once(f"Right encoder resolution: {data.resolution}")
+        rospy.loginfo_once(f"Right encoder type: {data.type}")
+        # store data value
+        self._ticks_right = data.data
+
+    def run(self):
+        # publish received tick messages every 0.05 second (20 Hz)
+        rate = rospy.Rate(20)
+        while not rospy.is_shutdown():
+            if self._ticks_right is not None and self._ticks_left is not None:
+                # start printing values when received from both encoders
+                msg = f"Wheel encoder ticks [LEFT, RIGHT]: {self._ticks_left}, {self._ticks_right}"
+                rospy.loginfo(msg)
+            rate.sleep()
+
+if __name__ == '__main__':
+    # create the node
+    node = WheelEncoderReaderNode(node_name='wheel_encoder_reader_node')
+    # run the timer in node
+    node.run()
+    # keep spinning
+    rospy.spin()
+
+```
+
+Again, we make our node executable,
+
+    chmod +x ./packages/my_package/src/wheel_encoder_reader_node.py
+
+
+
+## Define launcher
+
+Similarly to what we did in the section [](ros-sub-define-launcher), we create a new launcher file
+`./launchers/wheel-encoder-reader.sh` with the content,
+
+```shell
+#!/bin/bash
+
+source /environment.sh
+
+# initialize launch file
+dt-launchfile-init
+
+# launch subscriber
+rosrun my_package wheel_encoder_reader_node.py
+
+# wait for app to end
+dt-launchfile-join
+```
+
+Let us now re-compile our project using the command
+
+    dts devel build -f
+
+
+
+## Launch the node
+
+We are now ready to run our reader node,
+
+    dts devel run -R ROBOT_NAME -L wheel-encoder-reader
+
+The `resolution` and `type` values will be printed at the top of the logs once.
+
+Then in the console, the received left and right encoder data will be printed. Try:
+
+* spinning the left/right wheel, in both directions
+* driving the robot back and forth with the virtual joystick
```

Comment:
Link to the virtual joystick docs page. Use [this](https://docs.duckietown.com/daffy/devmanual-docs/language/format/index.html#cross-book-references) and not the absolute URL.